apps/shell/app.js

```diff
@@ -40,6 +45,74 @@ app.set('views', path.join(__dirname, 'views'));
 // Mount the routes at the base URI
 app.use(process.env.PASSENGER_BASE_URI || '/', router);
 
+var terminals = {
+    instances: {
+
+    },
+
+    create: function (host, dir) {
+        var cmd = 'ssh';
+        var args = dir ? [host, '-t', 'cd \'' + dir.replace(/\'/g, "'\\''") + '\' ; exec ${SHELL} -l'] : [host];
+        var uuid = currentId;
+
+        this.instances[uuid] = pty.spawn(cmd, args, {
+            name: 'xterm-256color',
+            cols: 80,
+            rows: 30
+        });
+        return uuid;
+    },
+
+    exists: function (uuid) {
+        if (uuid in instances) {
+            return true;
+        } else {
+            return false;
+        }
+    },
+
+    get: function (uuid) {
+        return this.instance[uuid];
+    },
+
+    attach: function (uuid, ws) {
+        var term = this.get(uuid);
```

Comment:
If the instance does not exist, term will be undefined. Then term.resume() will crash. We should handle this situation gracefully. Not sure what the best workflow is from the user's perspective. Can recreate if you have a session, then you disconnect for the day (instead of 5 minutes) and come back and try to reconnect.

Reply:
My assessment was incorrect. If there are no instances, we would create a new instance on websocket connection and attach to that. Not sure what case we would end up with term being undefined...